drivers/net/nsos_sockets.c

```diff
@@ -0,0 +1,947 @@
+				  addr_mid, addrlen_mid);
+	if (ret < 0) {
+		return -errno_from_nsos_mid(-ret);
+	}
+
+	return ret;
+}
+
+static ssize_t nsos_recvfrom(void *obj, void *buf, size_t len, int flags,
+			     struct sockaddr *addr, socklen_t *addrlen)
+{
+	struct nsos_socket *sock = obj;
+	struct nsos_mid_sockaddr_storage addr_storage_mid;
+	struct nsos_mid_sockaddr *addr_mid = (struct nsos_mid_sockaddr *)&addr_storage_mid;
+	size_t addrlen_mid = sizeof(addr_storage_mid);
+	int flags_mid;
+	int ret;
+
+	ret = socket_flags_to_nsos_mid(flags);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	flags_mid = ret;
+
+	ret = nsos_recvfrom_with_poll(sock, buf, len, flags_mid,
+				      addr_mid, &addrlen_mid);
+	if (ret < 0) {
+		goto return_ret;
+	}
+
+	sockaddr_from_nsos_mid(addr, addrlen, addr_mid, addrlen_mid);
+
+return_ret:
+	if (ret < 0) {
+		errno = errno_from_nsos_mid(-ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+static const struct socket_op_vtable nsos_socket_fd_op_vtable = {
+	.fd_vtable = {
+		.read = nsos_read,
+		.write = nsos_write,
+		.close = nsos_close,
+		.ioctl = nsos_ioctl,
+	},
+	.bind = nsos_bind,
+	.connect = nsos_connect,
+	.listen = nsos_listen,
+	.accept = nsos_accept,
+	.sendto = nsos_sendto,
+	.sendmsg = nsos_sendmsg,
+	.recvfrom = nsos_recvfrom,
+};
+
+static bool nsos_is_supported(int family, int type, int proto)
+{
+	int dummy;
+	int err;
+
+	err = socket_family_to_nsos_mid(family, &dummy);
+	if (err) {
+		return false;
+	}
+
+	err = socket_type_to_nsos_mid(type, &dummy);
+	if (err) {
+		return false;
+	}
+
+	err = socket_proto_to_nsos_mid(proto, &dummy);
+	if (err) {
+		return false;
+	}
+
+	return true;
+}
+
+NET_SOCKET_OFFLOAD_REGISTER(nsos, CONFIG_NET_SOCKETS_OFFLOAD_PRIORITY, AF_UNSPEC,
+			    nsos_is_supported, nsos_socket_create);
+
+struct zsock_addrinfo_wrap {
+	struct zsock_addrinfo addrinfo;
+	struct sockaddr_storage addr_storage;
+	struct nsos_mid_addrinfo *addrinfo_mid;
+};
+
+/*
+ * (Zephyr)
+ * zsock_addrinfo_wrap
+ * -----------------------
+ * | zsock_addrinfo      |
+ * -----------------------    (trampoline)
+ * | sockaddr_storage    |    nsos_addrinfo_wrap
+ * -----------------------    -----------------------------
+ * | nsos_mid_addrinfo * | -> | nsos_mid_addrinfo         |
+ * -----------------------    -----------------------------
+ *                            | nsos_mid_sockaddr_storage |
+ *                            -----------------------------    (Linux host)
+ *                            | addrinfo *                | -> addrinfo
+ *                            -----------------------------
+ */
+
+static int addrinfo_from_nsos_mid(struct nsos_mid_addrinfo *nsos_res,
+				  struct zsock_addrinfo **res)
+{
+	struct zsock_addrinfo_wrap *res_wraps;
+	size_t idx_res = 0;
+	size_t n_res = 0;
+
+	for (struct nsos_mid_addrinfo *res_p = nsos_res; res_p; res_p = res_p->ai_next) {
+		n_res++;
+	}
+
+	if (n_res == 0) {
+		return 0;
+	}
+
+	res_wraps = calloc(n_res, sizeof(*res_wraps));
```

Comment:
This was meant to use embedded libc on Zephyr side. However since I used `k_malloc()` in other place, it makes sense to use `k_calloc()` here as well. Also replaced `free()` with `k_free()`.